---
title: Reanalysis of _Arabidposis_ mutation accumulation data
output: pdf_document
bibliography: athal.bib
cls: plos.csl
author: David Winter
---

#Introduction

This document describes the re-analysis of sequencing data from a mutation
accumulation experiment in _Arabidopsis thaliana_ [@ossowski_2010; @becker_2011]. 
Prior to running the analyses in this file, sequencing reads where
downloaded from SRA and aligned to the TAIR10 reference genome. We then ran
`accuMUlate` to identify putative mutations. 

A complete record of the steps required to generate theese mutation calls is
provided as [a github repository](https://github.com/dwinter/accumulate-demonstration/). 
The output files we used in this analysis are:

 * `accu_raw.out`: Data from every site for which `accuMUlate` estimated a
    muation probability > 0.1.
 *  `denom.out` A file containing the data on the number of A, C, G and T bases
   that could be called for a mutation if one was present. This file is produced 
   by the `denominate` executable that is shipped with `accuMUlate`


# The data

We begin by defining two helper functions that will aid in processing this data.

```{r}
is_het <- function(g) substr(g,1,1) != substr(g, 2,2)

read_accu <- function(file, ...){
    mu <- read.table(file, stringsAsFactors=FALSE, ...)
    names(mu) <- c("chrom", "from", "to", "ref", "mutant_sample", "mutation", 
                   "p_any", "p_one", "p_geno", "likelihood", "DP", 
                   "mutant_F", "mutant_R", "anc_in_mutant", "mutant_in_anc",
                   "MQ_diff", "insert_diff", "strand_pval", "pair_pval")
    mu
}
```

We then  read in the mutation data. It is helpful to remove to sites that
`accuMUlate` predicts to have been heterozygous in the ancestor of this experiment.
Due to the nature of this experiment (with lines maintained by selfing and no
sequencing data from the ancestral strain) these sites are likely to be
ancestral heterozygosities in which one allele drifted to fixation in 
a particular lineage. 

```{r}
mu <- read_accu("results/accu_raw.out")
genotypes <-  strsplit(mu$mutation, "->")
mu$desc_het <- sapply(genotypes, function(x) is_het(x[2]))
anc_het <- sapply(genotypes, function(x) is_het(x[1]))
mu <- mu[!anc_het,]
nrow(mu)
```

`accMUlate` provides the user with the most probable ancestral and descendant
genotypes at each putatively mutant site. Here we convert these diploid
genotypes changes to one of six possible base-pair mutations.

```{r}
mu_tab <- read.csv("mu_table.csv", stringsAsFactors=FALSE)
mu_lookup <- mu_tab[,2]
names(mu_lookup) <- mu_tab[,1]
mu$six_mus <- mu_lookup[mu$mutation]
table(mu$six_mus)
```

#accMUlate successfully identifies all validated mutations

This dataset provides a number of useful ways to test the ability of `accuMUlate`
to identify mutations. Two of these lines were analysed by Ossowski and colleagues, -@ossowski_2010, 
and the locations of validated mutations are thus available from the [`smal`
database](cefg.uestc.edu.cn/smal/viewe.php?sp=ATHA1001). Thus we can compare
`accuMUlate` results directly with published results. We start by reading in the 
published data from `smal` and creating subsets of the `accuMUlate` and
published results that include only the two lines they share in common.

```{r}
published <- read.csv("smal_mus.csv", header=0, stringsAsFactors=FALSE)

restudied_con <- published[published[,1] %in%  c(59, 119),]
restudied_acc <- mu[mu$mutant_sample %in% c("ERX386699", "ERX386705"),]
```
To compare the presence of absence of mutations in each dataset we add a new
column `loc` that encodes the position of a mutation in the TAIR10 genome
(`chrom:pos` in 1-based coordinates). All of the mutations validated in the
published work are included the putative mutations identified by `accuMUlate`.


```{r}
restudied_con$loc <- paste0(restudied_con[,2], ":", restudied_con[,4])
restudied_acc$loc <- paste0(restudied_acc$chrom, ":", restudied_acc$to)
mean(restudied_con$loc %in% restudied_acc$loc)
```

In addition, there are a number of putative mutations that were not reported in the
published work but nevertheless generated relatively high mutation probabilities
using `accuMUlate`.

```{r}
sum(!(restudied_acc$loc %in% restudied_con$acc))
```

# Site summary data helps distinguish true positives from false positives

`accuMUlate` produces [a set of statistics for each putatively mutated site](https://github.com/dwinter/accuMUlate/wiki/Site-statistics).
For the most part, these statistics compare reads supporting a putative mutation
with those supporting an ancestral allele and can therefore be used to identify
artefacts of sequencing, library preparation or alignment-to-reference that may
have produced false positive mutation calls. 

In our dataset we have a set of putative mutations that we know to be true
mutations (having been validated by Ossowski et al, 2010) and another set that
contains at least some false positive sites (and possibly mutations missed in
the earlier paper). By comparing the distributions of these site statistics 
between each class of  putative mutations we can investigate how well they 
discriminate true positives from other sites.

Two of the statistics are best used together. The `mutant_F` and `mutant_R`
columns contain the number of reads mapping to the forward and reverse stand 
of the reference genome. Because indels or repeats up or downstream of a site
can create false positives, it can be useful to filter our sites that are not
supported by some number of FWD or REV oriented reads. To accommodate this
approach we have to create a new column that has the minimum of these two
statistics. 

With that column added, we can compare a subset of these statistics between each
set of putative mutations.

```{r, warning=FALSE, fig.cap="Comparison of quality-control statistics calculated from validated and and not-validated mutations."}
library(reshape2)
library(ggplot2)
theme_set(theme_bw())

restudied_acc$strand_min <- apply(restudied_acc[,c("mutant_F","mutant_R")], 1 ,min)
validated <- restudied_acc$loc %in% restudied_con$loc 
restudied_acc$published <- ifelse(validated, "published", "not published")
stats <- c("p_one", "MQ_diff", "insert_diff", "p_one", "strand_pval", "DP", "strand_min")
molten <- melt(restudied_acc,  measure.vars=stats)

ggplot(molten, aes(published, value, group=published)) + 
  geom_boxplot() +  scale_x_discrete("") + 
  facet_wrap(~variable, scales="free_y") +
  scale_y_log10("Statistic")
```

It's clear from the first of these graphs that the mutation probability
calculated by `accuMUlate` reflect the probability that putative mutation is
validated. All validated mutations generated a probability of $1.0$. The
non-validated set contains a number of sites with a mutation probability of
$1.0$ but also several with lower probabilities. There are also clears
differences in the statistics used to compare mutant and non-mutant reads. 
Notably, the not-published subset has relatively high values for `MQ_diff` and 
`insert_diff` (i.e. difference in mapping quality and inferred insert-size 
between ancestral and descendant alleles) and a number of samples with very high 
sequencing depth. 

With a larger set of mutations, and known false-positives, it would be possible
to use logistic regression to identify statistics that are particularity powerful
at distinguishing true from false positive mutations. Here we will simply use
these statistics to filter out sites. Because `MQ-diff` and `insert-diff` appear
to be informative we will use these statistics along with a requirement that
mutations be supported by $> 6$ reads in both Fwd and Rev orientation.
Only approximately one third of the mutative mutations fit these criteria.


```{r}
filt <- (mu$p_one > 0.1 & mu$mutant_F > 6 & mu$mutant_R > 6 
         & mu$MQ_diff < 5 & mu$insert_diff < 5)
mean(filt)
```

This filtering approach also removes the sites with very high depth.

```{r}
max(mu[filt,]$DP)
```

These results also contain a number of apparent multi-nucleotide mutations.
In our experience, apparent multi-nucleotide mutations are often artefacts
produced around indels. On the other hand, multi-nucleotide mutations certainly do
occur and if we were analysing this dataset completely we would Sanger sequence
these loci to investigate them further. For the purposes of this demonstration
we are most interested in the single nucleotide mutation rate so we will remove 
these sites. Here we remove any mutation that is within 2 bases of another 
mutation to  produce a "single nucleotide mutation" dataset.

```{r}
genomic_distance <- function(i,j, bed) {
    if(i == 0){
        return(NA)
    }
    if(i > nrow(bed)){
        return(NA)
    }
    if(bed[i,"chrom"] == bed[j, "chrom"]){
        return(abs(bed[i,"from"] - bed[j,"from"]))
    }
    return(-1)
}

good_mu <- mu[filt,]
good_mu$dist <- 0
for(i in 1:nrow(good_mu)){
    good_mu$dist[i] <- min(genomic_distance((i-1), i, good_mu), 
                           genomic_distance((i+1), i, good_mu), na.rm=TRUE)
}

snm <- good_mu[!(good_mu$dist %in% 1:2),]
```

These two filtering steps leave us with `r nrow(snm)` putative mutations.


# The mutation spectra is similar to previously published results. 

Both the mutations reported in the original paper and variants sampled from
natural populations of _A. thaliana_ show a very large excess of GC>AT
mutations. To calculate the rate of each mutation type in our data we need to
know the number of A,C, G and T bases from which a mutation could have been
called if one was present (i.e. the denominator of our mutation rate
calculation). The file `denom.out` contains an estimate of this number, produce
by simulating mutations and applying the filters described above.

`read_denom` is a helper function provided by `accumulate-tools` that reads
in denominator information. In this case we used a 600,000bp sample of the 
reference genome to estimate the proportion of call-able sites.

```{r}
source("accumulater.R")

genome_size <- sum(read.table("Athal.genome")[,2])
denom_by_base <- read_denom("results/denom.out", "denom_params.ini", genome_size, 6e6)
knitr::kable(denom_by_base)
```

Since the denominator is the number of sites multiplied by the number of generations
of MA we also need to know how many generations each of these lines evolved for
(`ngen`). With that information we can work out the rate of mutation for
each mutation-type.

```{r, fig.cap="Mutational spectrum"}
ngen <- rep(31,5)
nfreqs <-(denom_by_base %*% ngen)[,1]
denom_by_type <-  rep(c(sum(nfreqs[c(1,4)]), sum(nfreqs[c(2,3)])), each=3 )
mu_by_type <- as.data.frame(table(snm$six_mus, dnn="mutation_type"))
mu_by_type$rate <- mu_by_type$Freq/denom_by_type

ggplot(mu_by_type, aes(mutation_type, rate)) + 
    geom_bar(stat="identity") +
    scale_x_discrete("Mutation type") +
    scale_y_continuous("Mutation rate") +
    coord_flip()
```

# The overall estimate of mutation rate is very similar to  that previously reported

The first analysis of these MA lines produced an estimated mutation rate of 
$7 \times 10^{-9}$ mutations per site per generation. Using our filtered
data set we produce a very similar estimate:


```{r}
denom_by_sample <- colSums(denom_by_base) * ngen
nrow(snm) / sum(denom_by_sample)
```

# The mutation rate does not vary among samples

With this information we can estimate the mutation rate from each sample alone. 

```{r}
mu_by_sample <- table(snm$mutant_sample, dnn="sample")
sample_df <- data.frame(mu_by_sample, 
                        denom= denom_by_sample[names(mu_by_sample)])

sample_df$rate <- sample_df$Freq/sample_df$denom
ggplot(sample_df, aes(sample, rate)) + 
  geom_bar(stat="identity") +
  coord_flip()
```


# Mutations unique to accuMUlate have similar molecular spectra to validated mutations


## Heterozygotes

The analytical pipeline used by Ossowski et al (2010) was unable to identify
mutations that were still heterozygous (i.e. had not yet selfed to
homozygosity). The model used by `accuMUlate` can identify heterozygous
mutations, and indeed we find several such sites.

```{r}
table(snm$desc_het)
```

Heterozygous putative mutations also show a strong bias toward GC>AT mutations.

```{r}
table(snm[ snm$desc_het, "six_mus"])
```

## Mutations in the two lines analysed in both studies

Similarly, those putative mutations from the two samples re-analysed here that 
pass our filtering steps but were not reported by Ossowski et al (2010) show a
similar molecular spectrum, most notably an enrichment of GC>AT mutations.

```{r}
snm_restudy <- snm[snm$mutant_sample %in%  c("ERX386699", "ERX386705"),]
snm_loc <- paste0(snm_restudy$chrom, ":", snm_restudy$to) 
snm_restudy$pub <- snm_loc %in% restudied_con$loc
comp <- aggregate(as.factor(six_mus) ~ pub, data=snm_restudy, FUN=table)[,2]
rownames(comp) <- c("not published", "published")
knitr::kable(comp)
```

These spectra are not significantly different from each other.

```{r}
fisher.test(comp)
```
\pagebreak

# Summary tables

Finally, let's collect some of this data together to make the sort of tables you
might include in a scientific paper on an MA experiment. I have written some
functions to make this relatively straightforward. First, we can create a
summary table with information broken down by sample. 
```{r}

summary_table <- function(filtered, unfiltered, denom_by_base, ngen){
    denom_by_sample <- colSums(denom_by_base) * ngen
    sample_names <- names(denom_by_sample)
    rate_by_sample <- round(mu_by_sample/denom_by_sample,12)[sample_names]
    res <- data.frame(      
      sample = c(sample_names, "All hets", "All"),
      ngen = c(ngen, rep(sum(ngen),2)),
      n.initial = c(table(mu$mutant_sample)[sample_names],sum(mu$desc_het), nrow(mu)),
      n.filter = c(mu_by_sample[sample_names], sum(snm$desc_het), nrow(snm)),
      rate = c(rate_by_sample, "-", round(nrow(snm) / sum(denom_by_sample), 12))
    )
    res
}

summ_table <- summary_table(snm, mu, denom_by_base, rep(31,5))
knitr::kable(summ_table, caption="Summary of MA analysis by line")

```

We can also calculate the molecular spectrum of mutation by sample. Again, this
process is wrapped up in helper functions that may be helpful for other studies. 

```{r}

spectrum_denom <- function(denom_by_base, ngen){
    nsamp <- ncol(denom_by_base)
    snames <- colnames(denom_by_base)
    at_by_sample <- colSums(denom_by_base[c(1,4),] ) * ngen
    cg_by_sample <- colSums(denom_by_base[2:3,] ) * ngen
    res <- lapply(snames, 
                  function(x) rep(c(at_by_sample[x], cg_by_sample[x]),each=3))
    names(res) <- snames
    res
}

rate_table <- function(mu, denom_by_base, ngen, total=TRUE){
    snames <- colnames(denom_by_base)
    sample_spectrum <- aggregate(as.factor(six_mus) ~ mutant_sample, 
                                 data=mu, FUN=table, simplify=FALSE)
    spec_list <- structure(sample_spectrum[[2]], .Names=sample_spectrum$mutant_sample)
    D <- spectrum_denom(denom_by_base, ngen)
    res <- as.data.frame(sapply(snames, function(x) spec_list[[x]]/D[[x]]))
    if(total){
        total_denom <- rep(rowSums(sapply(D, "[", c(1,4))), each=3)
        total_n <- rowSums(sapply(spec_list, "+"))
        res$total <- total_n/total_denom
    }
    data.frame(t(res))
}

rate_spectrum <- rate_table(snm, denom_by_base, 31)
knitr::kable(formatC(as.matrix(rate_spectrum)), 
             caption="Mutational spectra ($\\hat{\\mu}$) by line")
```
 

# References
